content/2017.02.handling-python-projects-dependencies.rst

```diff
@@ -0,0 +1,225 @@
+release and it will install itself automatically with the right
+dependencies.
+
+Instead use your ``requirements.txt`` to set a know working set of your
+dependencies at the time of a release.
+
+We usually add a `build-requirements` makefile rule on our projects:
+
+.. code-block:: Makefile
+
+    VIRTUALENV = virtualenv --python=python3
+    
+    
+    build-requirements:
+    	$(eval TEMPDIR := $(shell mktemp -d))
+    	$(VIRTUALENV) $(TEMPDIR)
+    	$(TEMPDIR)/bin/pip install -U pip  # Upgrade pip
+    	$(TEMPDIR)/bin/pip install -Ue .   # Develop the current project locally
+    	# Freeze the dependencies ignoring the dependency links.
+    	$(TEMPDIR)/bin/pip freeze | grep -v -- '^-e' > requirements.txt
+
+If you run `make build-requirements` before tagging your release you
+will document what was the known working dependency set at the time of
+the release.
+
+Then you can install your program later using this file as a
+`dependency constrain file <https://pip.pypa.io/en/stable/user_guide/#constraints-files>`_:
+
+.. code-block:: console
+
+    pip install saucisson -c requirements.txt
+
+
+Avoid pining version in your setup.py
+=====================================
+
+You might be tempted to pin dependencies version in your setup.py
+
+Because you are using a contrain file you are safe for future updates
+that might break your code.
+
+On your CI not using a contrain file will help you to detect that you
+need to take some actions to support the new released version.
+
+There are some cases when you still want to pin some version:
+
+
+If you know that your project will not work with a lower version
+----------------------------------------------------------------
+
+If you are using a feature or API that didn't existed before or you
+had a bug that was fixed after.
+
+.. code-block::
+
+   psycopg > 2.5
+   colander >= 1.3.2
+   ujson >= 1.35
+
+
+If you know that your project doesn't support yet the next release
+------------------------------------------------------------------
+
+**I insist that this implies that the release already exists and that
+your test suite doesn't run on it.**
+
+In that case, and only in that case, you can do it for the shorted
+possible time until you ported your project to it.
+
+It is often the case for frameworks:
+
+.. code-block::
+
+   Pyramid < 1.8
+   django >1.6,<= 1.8
+
+The danger of doing it is that you migh create conflict and in that
+case Python will not let you start your application.
+
+An easy way to break things is to ping a max ``requests`` version for instance:
+
+.. code-block::
+
+   requests < 2.13
+
+If you do that you will end up having a lot of
+``pkg_resources.VersionConflict`` error when running your program
+because Python will check the dependency and will not accept to run if
+you have a greater ``requests`` version which can happen if another
+dependency already needed ``requests`` as a dependency and pip already
+installed it.
```

Comment:
`a lot of ... errors` and I think `refuse to run` is clearer. (I don't like `accept to run` but can't figure out how to fix it.)

Also, this feels kind of like a run-on sentence. By the time I got to `will not accept to run`, I already thought too much was happening here. I think you could break it into three: `If you do that, ... running your program`, `What's happening is that Python is checking ... version`, and `This can happen if ... installed it`.